### PR TITLE
GFORMS-2746 - Add ability to test for duplicates in Add To List component

### PR DESCRIPTION
**Pull request by cmanson**

### CODE REVIEW

File: app/uk/gov/hmrc/gform/eval/BooleanExprEval.scala [hunk 1]
```diff
@@ -147,6 +155,45 @@ class BooleanExprEval[F[_]: Monad] {
...
+        case (Atomic(component, atom), variadicValue) =>
+          (
+            component.maybeIndex,
+            component.baseComponentId.value,
+            Some(atom.value),
+            variadicValue.toSeq.map(canonicalStr(_, component.baseComponentId.value))
+          )
+      }
+      .groupBy(_._1)
+      .map(tuple => (tuple._1, tuple._2.map(tupleInner => (tupleInner._2, tupleInner._3, tupleInner._4)).toSet))
```
Comment (josef-vlach) [inline app/uk/gov/hmrc/gform/eval/BooleanExprEval.scala:192]: Better not to overly use `_n` accessors. What about this alternative?
```suggestion
      .view
      .mapValues { values =>
        values.map { case (_, baseComponentId, atom, canonicals) => (baseComponentId, atom, canonicals) }.toSet
      }
```

File: app/uk/gov/hmrc/gform/eval/BooleanExprEval.scala [hunk 2]
```diff
@@ -147,6 +155,45 @@ class BooleanExprEval[F[_]: Monad] {
...
+      .values
+
+    compare.size != compare.toSet.size
```
Comment (josef-vlach) [inline app/uk/gov/hmrc/gform/eval/BooleanExprEval.scala:195]: Let's use typesafe comparison `=!=`

**Review by josef-vlach — CHANGES_REQUESTED** [review]

File: app/uk/gov/hmrc/gform/eval/BooleanExprEval.scala (hunk 1 above)
Comment (cmanson) [reply app/uk/gov/hmrc/gform/eval/BooleanExprEval.scala]: Have made the requested changes

File: app/uk/gov/hmrc/gform/eval/BooleanExprEval.scala (hunk 2 above)
Comment (cmanson) [reply app/uk/gov/hmrc/gform/eval/BooleanExprEval.scala]: Yes that's done now too.

**Review by josef-vlach — APPROVED** [review]
